corehq/apps/data_interfaces/templates/data_interfaces/partials/auto_update_rule_list.html

```diff
@@ -9,6 +9,21 @@
     </button>
   </a>
 
+  <p class="pull-right" id="lock-container" data-bind="if: hasLinkedModels">
+    <span data-bind="ifnot: unlockLinkedData">
+      <button class="btn btn-primary" data-toggle="modal" data-target="#edit-warning-modal">
+        <i class="fa fa-link"></i>
+        {% trans 'Unlock Linked Rules For Me' %}
+      </button>
+    </span>
+    <span data-bind="if: unlockLinkedData">
```

Comment:
hidden element rather than `span`

Reply:
No issues changing them, but is there a rule we have for when to use one vs the other?

Reply:
I went over it a little bit [here](https://github.com/dimagi/commcare-hq/pull/33114#discussion_r1230945040)

but in general, `span`s should not wrap block elements, like `buttons`. In fact there's very little `span`s should wrap. Also, it adds unnecessary elements to the DOM, which when rendered provides no structural, semantic and stylistic purpose. the clutter adds up over time and should be avoided, and it some cases can seriously negatively impact performance of the page when used in excess.

Reply:
[This is a good article](https://blog.hubspot.com/website/span-vs-div) as well.

Reply:
Thanks for the description. The article you linked is discussing spans vs divs, but I was curious about when to use either spans/divs vs virtual elements. Because, to me, I share some of the feelings mentioned in [Reconsider Virtual Elements](https://github.com/knockout/knockout/issues/367) -- namely that many programmers view comments as comments and have a habit of skipping over them. The knockout docs don't do a good job explaining their use, with the main justification seemingly being that they allow a developer to handle HTML elements that cannot support random child elements otherwise (such as `<ul>` needing `<li>` items, not container elements). 

It might just be due to their simplicity, but none of the examples use virtual elements, even [Animated Transitions](https://knockoutjs.com/examples/animatedTransitions.html), where you could argue the template loop's div provides no value. [This stackoverflow response](https://stackoverflow.com/questions/20256355/is-there-anything-wrong-with-using-loads-of-virtual-elements-with-knockout) also seems to indicate, if anything, virtual elements are a slight performance overhead to normal elements.

Again, I don't have an issue moving to virtual elements here, just wanted to understand the rationale, since my searches aren't turning up much.

Reply:
Also found [another stackoverflow post](https://stackoverflow.com/questions/32023946/knockoutjs-virtual-element) mentioning that virtual elements cannot be used inside `<select>` elements for IE (if that's a relevant concern anymore).